src/main/kotlin/org/rust/lang/core/types/infer/TyLowering.kt

```diff
@@ -0,0 +1,338 @@
+                    } else {
+                        TyUnit.INSTANCE
+                    }
+                }
+                RsPsiSubstitution.Value.RequiredAbsent -> TyUnknown
+            }
+            paramTy to valueTy
+        }
+
+        val regionSubst = psiSubstitution.regionSubst.entries.mapNotNull { (psiParam, psiValue) ->
+            val param = ReEarlyBound(psiParam)
+            val value = if (psiValue is RsPsiSubstitution.Value.Present) {
+                psiValue.value.resolve()
+            } else {
+                return@mapNotNull null
+            }
+
+            param to value
+        }.toMap()
+
+        val constSubst = psiSubstitution.constSubst.entries.associate { (psiParam, psiValue) ->
+            val param = CtConstParameter(psiParam)
+            val value = when (psiValue) {
+                RsPsiSubstitution.Value.OptionalAbsent -> param
+                RsPsiSubstitution.Value.RequiredAbsent -> CtUnknown
+                is RsPsiSubstitution.Value.Present -> {
+                    val expectedTy = psiParam.typeReference?.normType ?: TyUnknown
+                    psiValue.value.toConst(expectedTy, resolver)
+                }
+                is RsPsiSubstitution.Value.DefaultValue -> {
+                    val expectedTy = psiParam.typeReference?.normType ?: TyUnknown
+                    psiValue.value.toConst(expectedTy, resolver)
+                }
+            }
+
+            param to value
+        }
+
+        val newSubst = Substitution(typeSubst, regionSubst, constSubst)
+
+        val assoc = if (withAssoc) {
+            psiSubstitution.assoc.mapValues {
+                when (val value = it.value) {
+                    is RsPsiSubstitution.AssocValue.Present -> lowerTy(value.value, null)
+                    RsPsiSubstitution.AssocValue.FnSugarImplicitRet -> TyUnit.INSTANCE
+                }
+            }
+        } else {
+            emptyMap()
+        }
+
+        return BoundElement(element, subst + newSubst, assoc)
+    }
+
+    companion object {
+        private fun withoutContext(): TyLowering = TyLowering(emptyMap())
+
+        fun lowerTypeReference(type: RsTypeReference): Ty {
+            val resolvedNestedPaths = if (type is RsBaseType) {
+                val kind = type.kind
+                if (kind is RsBaseTypeKind.Path) {
+                    RsPathReferenceImpl.resolveAllNestedPaths(kind.path)
+                } else {
+                    emptyMap()
+                }
+            } else {
+                emptyMap()
+            }
+            return TyLowering(resolvedNestedPaths).lowerTy(type, null)
+        }
+
+        fun <T : RsElement> lowerPathGenerics(
+            path: RsPath,
+            element: T,
+            subst: Substitution,
+            resolver: PathExprResolver,
+        ): BoundElement<T> {
+            return withoutContext().instantiatePathGenerics(path, element, subst, resolver, withAssoc = true)
```

Comment:
Why we use `withAssoc = true` here but `withAssoc = target is RsTraitItem` in `lowerTy`?

Reply:
`assoc` was previously unused, so in the case of `target !is RsTraitItem`, so I disabled its calculation.
![image](https://user-images.githubusercontent.com/3221931/189696931-941c4a57-dad9-4c88-8781-bc274193040d.png)